app/models/concerns/data_uri.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+# This module provides a way to generate a data URI for a file or URL.
+# It encodes the file content or the content fetched from a URL into a Base64 string
+# and formats it as a data URI. This can be useful for embedding files directly in HTML
+# or other contexts where a data URI is needed.
+#
+# Example usage:
+# class Media < ApplicationRecord
+#   include DataUri
+# end
+#
+# The `data_uri` attribute should be defined in the model's database schema.
+# Ensure that the `data_uri` column is added to the model's database table.
+require_relative '../../../lib/data_uri_builder'
+module DataUri
+  extend ActiveSupport::Concern
+
+  included do
+    before_save :generate_data_uri, if: :should_generate_data_uri?
+  end
+
+  def should_generate_data_uri?
+    file_changed = if self[:checksum].nil?
+                     return true
+                   else
+                     self[:checksum] != checksum
+                   end
+    url_changed = respond_to?(:will_save_change_to_url?) ? will_save_change_to_url? : false
+    file_changed || url_changed
+  end
+
+
+  def previous_checksum
+    @previous_checksum ||= self[:file_checksum]
+  end
+
+  def current_file_checksum
+    file.attachment&.blob&.checksum
+  end
+
+  def generate_data_uri
```

Comment:
I'm not sure about this approach. It means storing the file in the database as well as on disk, but with the 30% bloat that comes with base64 encoding. What is the goal? What is the benefit?

Reply:
If you want to return the data URIs to the frontend, I would consider eager-loading the ActiveStorage attachment with blob, and base64 encoding it with the output. I would also consider caching each one to disk (such as in tmp/cache) if it is a performance concern to do the base64 on demand. But storing the data URI on disk, not so much.

Reply:
The idea here is, in modern computing, storage is cheap, but cpu and memory cycles are expensive overhead. This preload the URI, then when the reply is returned you aren't building on the fly. Which isn't a concern when you have 1 or 2, but if you have 20 attachments, the overhead on building those on the fly can cause significant lag in API response, severely impacting UX. Additionally, the 1-2 sec response of the ActiveStorage controller on providing the blob at runtime for each picture viewed hampered UX as well, So the hit is taken on the front end on generation, when the slight increase in processing time is barely noticeable, since these are done in discrete transactions, and the reward on the backend is near instant display of media.

Reply:
In production, I am not seeing a significant size increase from this.  I am, however, seeing that `versions`, specifically the `object` field, does drastically affect size:

```sql
SELECT tablename,
	pg_size_pretty(pg_relation_size(tablename::regclass)) AS data_size,
	pg_size_pretty(pg_database_size('historyforge')) as db_size,
	to_char(pg_relation_size(tablename::regclass)*100.0/pg_database_size('historyforge'), 'FM999.00%') as percentage_of_db_size
FROM 
	pg_tables
WHERE
	schemaname = 'public'
ORDER BY
	pg_relation_size(tablename::regclass) DESC
LIMIT 
	1;
```

| tablename | data_size | db_size | percentage_of_db_size
|-----------|-----------|---------|-----------------------
| versions | 84.8 MB | 119 MB | 71.26% |